### ISSUE TITLE
instance stuck in orphan mode after restart

### ISSUE THREAD

**Issue by timjohncurtis** [tarantool/cartridge#746]
Unsure as this is a cartridge issue or a base tarantool issue. I am using the latest 2.2 release `2.2.2-72-g7869d5e` When trying to roll a deployment by restarting the nodes in sequence and reconciling the cluster in between, I end up with the instance in `OperationError` state which is not recoverable from, as this instance is stuck in Orphan mode

I am running in kubernetes, using the tarantool-operator which seems to be working as expected other than this issue, which seems to be in cartridge itself

Logs from dead node
```
Membership encryption enabled
Could not reach node: api-0-1.userscripting.default.svc.cluster.local:3301 - suspect
Refuting the rumor that we are suspect
ttool_address: 10.3.130.96:3301
reg_address: 10.3.130.96:8081
init_slack_and_nats_communication
slack initialised
nats initialised
registered ttool-userscripting-10.3.130.96-8081
Using advertise_uri "api-0-1.userscripting.default.svc.cluster.local:3301"
Membership encryption enabled
Membership BROADCAST sent to 127.0.0.1:3302
Membership BROADCAST sent to 10.3.130.96:3302
Membership BROADCAST sent to 127.0.0.1:3301
Membership BROADCAST sent to 10.3.130.96:3301
Membership BROADCAST sent to 127.0.0.1:3300
Membership BROADCAST sent to 10.3.130.96:3300
started
Listening HTTP on 0.0.0.0:8081
started
Remote control listening on 0.0.0.0:3301
Instance state changed:  -> Unconfigured
entering the event loop
Refuting the rumor that we are dead
Could not reach node: core-0-1.userscripting.default.svc.cluster.local:3301 - suspect
Could not reach node: core-0-2.userscripting.default.svc.cluster.local:3301 - suspect
Refuting the rumor that we are suspect
Could not reach node: api-0-2.userscripting.default.svc.cluster.local:3301 - suspect
Node timed out: core-0-1.userscripting.default.svc.cluster.local:3301 - dead
Node timed out: core-0-2.userscripting.default.svc.cluster.local:3301 - dead
Refuting the rumor that we are suspect
Node timed out: api-0-2.userscripting.default.svc.cluster.local:3301 - dead
Peer closed
Peer closed
Peer closed
Peer closed
Peer closed
Instance state changed: Unconfigured -> BootstrappingBox
Calling box.cfg()...
Creating configuration file: /etc/tarantool/config.yml
Config:
---
pid_file: /var/run/tarantool/tarantool.pid
instance_uuid: 71d9a62b-204b-5028-ba00-d78573d60c6b
memtx_memory: 268435456
vinyl_dir: /opt/tarantool/tmp/db
custom_proc_title: lapti@api-0-1
read_only: true
memtx_dir: /opt/tarantool/tmp/db
replicaset_uuid: 7e2d78b9-fa67-56e8-a935-5b1ef0a46b42
listen: null
work_dir: /opt/tarantool
replication:
- xxxx:xxxx@api-0-0.userscripting.default.svc.cluster.local:3301
wal_dir: /opt/tarantool/tmp/db
...

2020-04-06 14:35:33.169 [1] main/160/remote_control/10.3.147.147:4103 C> Tarantool 2.2.2-72-g7869d5e
2020-04-06 14:35:33.169 [1] main/160/remote_control/10.3.147.147:4103 C> log level 5
2020-04-06 14:35:33.169 [1] main/160/remote_control/10.3.147.147:4103 I> mapping 268435456 bytes for memtx tuple arena...
2020-04-06 14:35:33.169 [1] main/160/remote_control/10.3.147.147:4103 I> mapping 134217728 bytes for vinyl tuple arena...
2020-04-06 14:35:33.170 [1] main/160/remote_control/10.3.147.147:4103 I> instance uuid 71d9a62b-204b-5028-ba00-d78573d60c6b
2020-04-06 14:35:33.171 [1] main/160/remote_control/10.3.147.147:4103 I> connecting to 1 replicas
2020-04-06 14:35:33.173 [1] main/167/applier/admin@api-0-0.userscrip I> remote master eaa81656-5977-5309-b2d3-5c6764a6d813 at 10.3.149.209:3301 running Tarantool 2.2.2
2020-04-06 14:35:33.174 [1] main/160/remote_control/10.3.147.147:4103 I> connected to 1 replicas
2020-04-06 14:35:33.175 [1] main/167/applier/admin@api-0-0.userscrip I> authenticated
2020-04-06 14:35:33.175 [1] main/160/remote_control/10.3.147.147:4103 I> bootstrapping replica from eaa81656-5977-5309-b2d3-5c6764a6d813 at 10.3.149.209:3301
2020-04-06 14:35:33.176 [1] main/167/applier/admin@api-0-0.userscrip I> cluster uuid 7e2d78b9-fa67-56e8-a935-5b1ef0a46b42
2020-04-06 14:35:33.183 [1] main/167/applier/admin@api-0-0.userscrip I> assigned id 1 to replica eaa81656-5977-5309-b2d3-5c6764a6d813
2020-04-06 14:35:33.183 [1] main/167/applier/admin@api-0-0.userscrip I> assigned id 2 to replica 71d9a62b-204b-5028-ba00-d78573d60c6b
2020-04-06 14:35:33.183 [1] main/167/applier/admin@api-0-0.userscrip I> initial data received
2020-04-06 14:35:33.183 [1] main/167/applier/admin@api-0-0.userscrip I> final data received
2020-04-06 14:35:33.183 [1] snapshot/101/main I> saving snapshot `/opt/tarantool/tmp/db/00000000000000000014.snap.inprogress'
2020-04-06 14:35:33.185 [1] snapshot/101/main I> done
2020-04-06 14:35:33.186 [1] main/160/remote_control/10.3.147.147:4103 I> ready to accept requests
2020-04-06 14:35:33.186 [1] main/160/remote_control/10.3.147.147:4103 I> synchronizing with 1 replicas
2020-04-06 14:35:33.187 [1] main/167/applier/admin@api-0-0.userscrip I> subscribed
2020-04-06 14:35:33.187 [1] main/167/applier/admin@api-0-0.userscrip I> remote vclock {1: 14} local vclock {1: 14}
2020-04-06 14:35:33.187 [1] main/168/applierw/admin@api-0-0.userscri C> leaving orphan mode
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 I> replica set sync complete
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 C> leaving orphan mode
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 I> set 'custom_proc_title' configuration option to "lapti@api-0-1"
2020-04-06 14:35:33.187 [1] main/162/checkpoint_daemon I> scheduled next checkpoint for Mon Apr  6 16:35:22 2020
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 I> set 'instance_uuid' configuration option to "71d9a62b-204b-5028-ba00-d78573d60c6b"
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 I> set 'read_only' configuration option to true
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 I> set 'replication' configuration option to ["admin@api-0-0.userscripting.default.svc.cluster.local:3301"]
2020-04-06 14:35:33.187 [1] main/160/remote_control/10.3.147.147:4103 I> set 'replicaset_uuid' configuration option to "7e2d78b9-fa67-56e8-a935-5b1ef0a46b42"
2020-04-06 14:35:33.188 [1] main/175/console/unix/:/var/run/tarantool I> started
2020-04-06 14:35:33.188 [1] main/160/remote_control/10.3.147.147:4103 I> Making sure user "admin" exists...
2020-04-06 14:35:33.188 [1] main/160/remote_control/10.3.147.147:4103 I> Granting replication permissions to "admin"...
2020-04-06 14:35:33.188 [1] main/160/remote_control/10.3.147.147:4103 I> Setting password for user "admin" ...
2020-04-06 14:35:33.188 [1] main/160/remote_control/10.3.147.147:4103 I> set 'read_only' configuration option to false
2020-04-06 14:35:33.188 [1] main/160/remote_control/10.3.147.147:4103 I> set 'read_only' configuration option to true
2020-04-06 14:35:33.189 [1] iproto/101/main I> binary: bound to 0.0.0.0:3301
2020-04-06 14:35:33.189 [1] main/112/remote_control/0.0.0.0:3301 utils.c:1018 E> LuajitError: builtin/socket.lua:83: attempt to use closed socket
2020-04-06 14:35:33.189 [1] main/160/remote_control/10.3.147.147:4103 I> set 'listen' configuration option to "3301"
2020-04-06 14:35:33.189 [1] main/160/remote_control/10.3.147.147:4103 I> Instance state changed: BootstrappingBox -> BoxConfigured
2020-04-06 14:35:33.189 [1] main/160/remote_control/10.3.147.147:4103 I> Instance state changed: BoxConfigured -> ConnectingFullmesh
2020-04-06 14:35:33.189 [1] main/160/remote_control/10.3.147.147:4103 I> connecting to 3 replicas
2020-04-06 14:35:33.190 [1] main/177/applier/admin@api-0-1.userscrip I> remote master 71d9a62b-204b-5028-ba00-d78573d60c6b at 10.3.130.96:3301 running Tarantool 2.2.2
2020-04-06 14:35:33.191 [1] main/178/applier/admin@api-0-2.userscrip I> remote master bf29f3cf-6936-5178-a829-b86a765ec6c8 at 10.3.129.218:3301 running Tarantool 2.2.2
2020-04-06 14:35:33.192 [1] main/176/applier/admin@api-0-0.userscrip I> remote master eaa81656-5977-5309-b2d3-5c6764a6d813 at 10.3.149.209:3301 running Tarantool 2.2.2
2020-04-06 14:35:33.193 [1] main/160/remote_control/10.3.147.147:4103 I> connected to 3 replicas
2020-04-06 14:35:33.193 [1] main/160/remote_control/10.3.147.147:4103 I> synchronizing with 3 replicas
2020-04-06 14:35:33.193 [1] main/178/applier/admin@api-0-2.userscrip I> authenticated
2020-04-06 14:35:33.193 [1] main/178/applier/admin@api-0-2.userscrip I> subscribed
2020-04-06 14:35:33.193 [1] main/178/applier/admin@api-0-2.userscrip I> remote vclock {1: 15, 2: 1, 3: 1} local vclock {1: 14, 2: 1}
2020-04-06 14:35:33.193 [1] main/176/applier/admin@api-0-0.userscrip I> authenticated
2020-04-06 14:35:33.194 [1] main/178/applier/admin@api-0-2.userscrip I> can't read row
2020-04-06 14:35:33.194 [1] main/178/applier/admin@api-0-2.userscrip xrow.c:1069 E> ER_PROC_LUA: Missing .xlog file between LSN 15 {1: 14, 2: 1} and 16 {1: 15, 2: 1}
2020-04-06 14:35:33.194 [1] main/160/remote_control/10.3.147.147:4103 C> failed to synchronize with 2 out of 3 replicas
2020-04-06 14:35:33.194 [1] main/160/remote_control/10.3.147.147:4103 C> entering orphan mode
2020-04-06 14:35:33.194 [1] main/160/remote_control/10.3.147.147:4103 I> set 'replication' configuration option to ["admin@api-0-0.userscripting.default.svc.cluster.local:3301","admin@api-0-1.userscripting.default.svc.cluster.local:3301","admin@api-0-2.userscripting.default.svc.cluster.local:3301"]
2020-04-06 14:35:33.194 [1] main/160/remote_control/10.3.147.147:4103 confapplier.lua:120 E> Instance entering failed state: ConnectingFullmesh -> OperationError
BoxError: Replication setup failed, instance in orphan mode
stack traceback:
	...rantool/.rocks/share/tarantool/cartridge/confapplier.lua:226: in function <...rantool/.rocks/share/tarantool/cartridge/confapplier.lua:202>
	[C]: in function 'xpcall'
	/opt/tarantool/.rocks/share/tarantool/errors.lua:139: in function </opt/tarantool/.rocks/share/tarantool/errors.lua:123>
	...tool/.rocks/share/tarantool/cartridge/remote-control.lua:89: in function <...tool/.rocks/share/tarantool/cartridge/remote-control.lua:62>
	[C]: in function 'pcall'
	...tool/.rocks/share/tarantool/cartridge/remote-control.lua:262: in function <...tool/.rocks/share/tarantool/cartridge/remote-control.lua:149>
	[C]: in function 'xpcall'
	/opt/tarantool/.rocks/share/tarantool/errors.lua:139: in function 'pcall'
	...tool/.rocks/share/tarantool/cartridge/remote-control.lua:311: in function <...tool/.rocks/share/tarantool/cartridge/remote-control.lua:295>
	[C]: in function 'pcall'
	builtin/socket.lua:1073: in function <builtin/socket.lua:1071>
2020-04-06 14:35:33.194 [1] main/176/applier/admin@api-0-0.userscrip I> subscribed
2020-04-06 14:35:33.194 [1] main/176/applier/admin@api-0-0.userscrip I> remote vclock {1: 14} local vclock {1: 14, 2: 1}
2020-04-06 14:35:40.326 [1] main/155/remote_control/10.3.147.147:4102 I> Peer closed
2020-04-06 14:35:40.326 [1] main/158/remote_control/10.3.147.147:4103 I> Peer closed
2020-04-06 14:35:40.327 [1] main/159/remote_control/10.3.147.147:4103 I> Peer closed
```

logs from instance which was using to configure: 
```4-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at api-0-0.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at api-0-2.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at core-0-0.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at core-0-1.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at core-0-2.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at core-0-3.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at core-0-4.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at userscripts-0-0.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at userscripts-0-1.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at userscripts-0-2.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at userscripts-0-3.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:331 W> Committed config at userscripts-0-4.userscripting.default.svc.cluster.local:3301
2020-04-06 14:35:33.195 [1] main/11633/http/127.0.0.1:54614 twophase.lua:337 E> Error committing config at api-0-1.userscripting.default.svc.cluster.local:3301:
BoxError: Replication setup failed, instance in orphan mode
stack traceback:
	...rantool/.rocks/share/tarantool/cartridge/confapplier.lua:226: in function <...rantool/.rocks/share/tarantool/cartridge/confapplier.lua:202>
	[C]: in function 'xpcall'
	/opt/tarantool/.rocks/share/tarantool/errors.lua:139: in function </opt/tarantool/.rocks/share/tarantool/errors.lua:123>
	...tool/.rocks/share/tarantool/cartridge/remote-control.lua:89: in function <...tool/.rocks/share/tarantool/cartridge/remote-control.lua:62>
	[C]: in function 'pcall'
	...tool/.rocks/share/tarantool/cartridge/remote-control.lua:262: in function <...tool/.rocks/share/tarantool/cartridge/remote-control.lua:149>
	[C]: in function 'xpcall'
	/opt/tarantool/.rocks/share/tarantool/errors.lua:139: in function 'pcall'
	...tool/.rocks/share/tarantool/cartridge/remote-control.lua:311: in function <...tool/.rocks/share/tarantool/cartridge/remote-control.lua:295>
	[C]: in function 'pcall'
	builtin/socket.lua:1073: in function <builtin/socket.lua:1071>
during net.box call to api-0-1.userscripting.default.svc.cluster.local:3301, function "_G.__cartridge_clusterwide_config_commit_2pc"
stack traceback:
	/opt/tarantool/.rocks/share/tarantool/cartridge/pool.lua:114: in function </opt/tarantool/.rocks/share/tarantool/cartridge/pool.lua:104>
```

**Comment by olegrok** [comment]
```
2020-04-06 14:35:33.194 [1] main/178/applier/admin@api-0-2.userscrip xrow.c:1069 E> ER_PROC_LUA: Missing .xlog file between LSN 15 {1: 14, 2: 1} and 16 {1: 15, 2: 1}
```

You have problems with replication. I'm not sure (but looks like) that it could be caused by https://github.com/tarantool/tarantool/issues/4759

**Comment by timjohncurtis** [comment]
I will close this issue, do you know if there is a 2.x tarantool branch that I could build to test if the fixes applied for that issue will fix my problem? This is a blocker for our deployment at the minute.

**Comment by olegrok** [comment]
Don't hurry to close the issue. Try to use this patch 2.2.2-48-gdb72782c7/2.3.1-52-g1d223d0d7/2.4.0-71-gdbcfaf70d or later.

Then try to drop xlogs on your replicas.

**Comment by timjohncurtis** [comment]
ok thank you for your help @olegrok I will try those today

**Comment by timjohncurtis** [comment]
I have cleared all xlog files and tried all 3 of those revisions and still this `OperationError` issue is occuring

**Comment by olegrok** [comment]
@rosik Could you help, please?

**Comment by rosik** [comment]
Hi. Could you tell us more about your scenario? I'm trying to reproduce it, and can't manage - how. From the log I see that the instance initially doesn't have a snapshot (so it's his first box.cfg), did you change the topology in between? Or where did the snapshot go?

Also note, that in cartridge 2.0.2 we prevented transition from `ConnectingFullmesh` to `OperationError` state, so it's recoverable now.

@vasiliy-t could you join our discussion, please.

**Comment by timjohncurtis** [comment]
Thanks for coming back to me so quickly, I found the issue was in our code after some time investigating. So We are using the tarantool-operator for kubernetes. The issue is entirely that our code does not honour the `TARANTOOL_WORK_DIR` environment variable for it's storage as we were overriding `workdir` in the `cartridge.cfg` call, I have corrected our code and the cluster is running great now. I will close the issue.
